### ISSUE TITLE
EventPool: Count Used Events and Dynamically Extend

### ISSUE THREAD

**Issue by anshumang** [ComputationalRadiationPhysics/picongpu#1189]
Referring to [this](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/libPMacc/include/eventSystem/Manager.tpp#L198).
## 

ideas (@ax3l):
- remember events that are in use (in-use counter in pool; "stolen" events with reference counter?)
- log warning and allocate more events in case the pool is empty

**Comment by psychocoderHPC** [comment]
This is a randomly chosen number which is large enouph for the most cases. 
The user code like `PIConGPU` can increase the number by calling `Environment<simDim>::Manager().getEventPool().addEvents( 1000 );

Small example why we need a lot of events:
One **PIConGPU** iteration of the `Kelvin-Helmholtz` example needs
  - ~110 events for Host2Device
  - ~265 events for Device2Host
  - ~228 events for Device2Device
  - ~500 events for kernel calls

**Comment by ax3l** [comment]
doesn't look that save to me, neither the number of elements that is reserved before scheduling the same event twice again (from the beginning) nor the way that it is completely unknown how many events are in use right now.

**Comment by psychocoderHPC** [comment]
True not nice and I have now a example were we can deadlock.
```
Events: e1
Streams; s1, s1

- record(e1 in s1)
- streamWait(s2 on e1)  //e1 is not finished at this moment
- record(e1 in s2) //e1 is not finished at this moment
//now we had a deadlock
```

**Comment by ax3l** [comment]
(note: current work-around is too keep the number of available events high)

**Comment by ax3l** [comment]
closed via #1631
